File: Count digit groupings of a number.cpp
class Solution{
	public:
	int TotalCount(string str){
	     int n = str.length();
        int presum[105] = {0};n = str.length();
	    for(int i=1;i<=n;i++)    presum[i] = presum[i-1] + (str[i-1]-'0');
	    
	    vector<vector<int>> dp(n+1, vector<int>(presum[n]+1, 0));
	    //base case
	    for(int prev=0;prev<=presum[n];prev++)    dp[n][prev] = 1;
	    
	    for(int ind=n-1;ind>=0;ind--){
	        for(int prev=presum[n]; prev >= 0;prev--){
	            
	            //just copying the recurrence
	            int res = 0;
	            for(int k=ind; k < n;k++){
	                int sum = presum[k+1] - presum[ind];
	                if(sum >= prev)     res += dp[k+1][sum];
	            }
	            dp[ind][prev] = res;
	        }
	    }
	    
	    return dp[0][0];

	}

};

File: Count digit groupings of a number.py


class Solution:
    def __init__(self):
        self.dp = []
        
	def TotalCount(self, s):
	    maxsum = (len(s) * (len(s) + 1)) // 2 * 9
	    self.dp = [[-1]* (maxsum + 1) for _ in range(len(s))]
	    return self.grouping(s,0,0)
	    
	def grouping(self,s,index,sum):
	    if index == len(s):
	        return 1
	   
	    if self.dp[index][sum] != -1:
	        return self.dp[index][sum]
	        
	    currSum = 0
	    ans = 0
	    
	    for i in range(index, len(s)):
	        currSum += int(s[i])
	        
	        if currSum >= sum:
	            ans += self.grouping(s,i+1,currSum)
	            
	    self.dp[index][sum] = ans
	    
	    return ans
